README.md
# dribbble-experiments

1) Experiments with Dribbble API
2) URI parsing and match library

## Build

    $ lein uberjar

## Run

Usage:

    lein run <Username> <AccessToken>

Example:

    $ lein run "louiethelowe" "960d368303253732c7d0d6e2081c5ab8efe1e79349a223ce4d8458be4e49948c"

Help:

    $ lein run help

## Jar Usage

Args are the same as in "Run" part

    $ java -jar dribbble-experiments-0.1.0-SNAPSHOT.jar [args]

## Run Tests for URI parsing and match library

    $ lein test

# URI parsing and match library

## Examples

```clojure
user=>(def pattern (pattern-info 
    "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type); fragment(?paragraph)"))

user=>(recognize 
    pattern 
    "https://dribbble.com/some-username/status/1905065-Travel-Icons-pack?list=users&offset=1&page=34#paragraph=3")
[   [:user "some-username"] 
    [:id "1905065-Travel-Icons-pack"] 
    [:offset "1"] 
    [:type "users"] 
    [:paragraph "paragraph=3"]]

user=>(recognize 
    pattern 
    "https://twitter.com/some-username/status/1905065-Travel-Icons-pack?list=users&offset=1&page=34#paragraph=3")
nil

user=>(def pattern-any-host (pattern-info 
    "host(); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type); fragment(?paragraph)"))

user=>(recognize 
    pattern-any-host 
    "https://twitter.com/some-username/status/1905065-Travel-Icons-pack?list=users&offset=1&page=34#paragraph=3")
[   [:user "some-username"] 
    [:id "1905065-Travel-Icons-pack"] 
    [:offset "1"] 
    [:type "users"] 
    [:paragraph "paragraph=3"]]
```

More examples can be seen in "test/dribbble_experiments/core_test.clj"

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-09-16
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-09-16
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/dribbble-experiments/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/dribbble-experiments/compare/0.1.0...0.1.1


test/dribbble_experiments/core_test.clj
(ns dribbble-experiments.core-test
  (:require [clojure.test :refer :all]
            [dribbble-experiments.core :refer :all]
            [dribbble-experiments.url_match :refer :all]))

(deftest url-match-tests
  (testing "URL-MATCH. Pattern Info."
    (is (=  (:fragment (pattern-info "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type); fragmmmment(?paragraph)"))
            nil))
    (is (=  (:fragment (pattern-info "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type); fragment(paragraph)"))
            nil))
    (is (=  (:fragment (pattern-info "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type)"))
            nil))
    (is (=  (:query (pattern-info "host(dribbble.com); path(?user/status/?id); querrrryparam(list=?type); fragment(?paragraph)"))
            nil))
    (is (=  (:query (pattern-info "host(dribbble.com); path(?user/status/?id); fragment(?paragraph)"))
            nil))
    (is (=  (:query (pattern-info "host(dribbble.com); path(?user/status/?id); queryparam(list=type); fragment(?paragraph)"))
            nil))
    (is (=  (:query (pattern-info "host(dribbble.com); path(?user/status/?id); queryparam(=?type); fragment(?paragraph)"))
            nil))
    (is (=  (:query (pattern-info "host(dribbble.com); path(?user/status/?id); queryparam(list=?); fragment(?paragraph)"))
            nil))
    (is (=  (:path (pattern-info "host(dribbble.com); queryparam(offset=?offset); queryparam(list=?type); fragment(?paragraph)"))
            nil))
    (is (=  (:path (pattern-info "host(dribbble.com); path(); queryparam(offset=?offset); queryparam(list=?type); fragment(?paragraph)"))
            nil))
    (is (=  (:host (pattern-info "host(http://dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type); fragment(?paragraph)"))
            nil))
    (is (=  (:host (pattern-info "path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type); fragment(?paragraph)"))
            nil))
  )
  (testing "URL-MATCH. Bad uri."
    (is (= nil (recognize 
                  (pattern-info "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type); fragment(?paragraph)") 
                  "https://twitter.com/some-username/status/1905065-Travel-Icons-pack?list=users&offset=1&page=34#paragraph=3")))
    (is (= nil (recognize 
                  (pattern-info "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type); fragment(?paragraph)") 
                  "https://dribbble.com/some-username/staaaaatus/1905065-Travel-Icons-pack?list=users&offset=1&page=34#paragraph=3")))
    (is (= nil (recognize 
                  (pattern-info "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type); fragment(?paragraph)") 
                  "https://dribbble.com/some-username/status?list=users&offset=1&page=34#paragraph=3")))
    (is (= nil (recognize 
                  (pattern-info "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type); fragment(?paragraph)") 
                  "https://dribbble.com/some-username/1905065-Travel-Icons-pack?list=users&offset=1&page=34#paragraph=3")))
    (is (= nil (recognize 
                  (pattern-info "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type); fragment(?paragraph)") 
                  "https://dribbble.com/some-username/status/1905065-Travel-Icons-pack?liiiiiist=users&offset=1&page=34#paragraph=3")))
    (is (= nil (recognize 
                  (pattern-info "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type); fragment(?paragraph)") 
                  "https://dribbble.com/some-username/status/1905065-Travel-Icons-pack?offset=1&page=34#paragraph=3")))
    (is (= nil (recognize 
                  (pattern-info "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type); fragment(?paragraph)") 
                  "https://dribbble.com/some-username/status/1905065-Travel-Icons-pack?list=users&offset=1&page=34")))
    (is (= nil (recognize 
                  (pattern-info "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type); fragment(?paragraph)") 
                  "https://dribbble.com/some-username/status/1905065-Travel-Icons-pack#paragraph=3")))
    (is (= nil (recognize 
                  (pattern-info "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type); fragment(?paragraph)") 
                  "https://dribbble.com/?list=users&offset=1&page=34#paragraph=3")))
    (is (= nil (recognize 
                  (pattern-info "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type); fragment(?paragraph)") 
                  "https://dribbble.com/some-username/status/1905065-Travel-Icons-pack")))
    (is (= nil (recognize 
                  (pattern-info "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type); fragment(?paragraph)") 
                  "https://dribbble.com/?list=users&offset=1&page=34")))
    (is (= nil (recognize 
                  (pattern-info "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type); fragment(?paragraph)") 
                  "https://dribbble.com/#paragraph=3")))
    (is (= nil (recognize 
                  (pattern-info "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type); fragment(?paragraph)") 
                  "https://dribbble.com/")))
    (is (= nil (recognize 
                  (pattern-info "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type); fragment(?paragraph)") 
                  "httttttttps://dribbble.com/some-username/status/1905065-Travel-Icons-pack?list=users&offset=1&page=34#paragraph=3")))
    (is (= nil (recognize 
                  (pattern-info "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type); fragment(?paragraph)") 
                  "://dribbble.com/some-username/status/1905065-Travel-Icons-pack?list=users&offset=1&page=34#paragraph=3")))
    (is (= nil (recognize 
                  (pattern-info "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type); fragment(?paragraph)") 
                  "dribbble.com/some-username/status/1905065-Travel-Icons-pack?list=users&offset=1&page=34#paragraph=3")))
  )
  (testing "URL-MATCH. Good uri."
    (let [inner-result (recognize 
                          (pattern-info "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type); fragment(?paragraph)") 
                          "https://dribbble.com/some-username/status/1905065-Travel-Icons-pack?list=users&offset=1&page=34#paragraph=3")]
      (is (= [[:user "some-username"] [:id "1905065-Travel-Icons-pack"] [:offset "1"] [:type "users"] [:paragraph "paragraph=3"]] inner-result)))
    (let [inner-result (recognize 
                          (pattern-info "host(); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type); fragment(?paragraph)") 
                          "https://dribbble.com/some-username/status/1905065-Travel-Icons-pack?list=users&offset=1&page=34#paragraph=3")]
      (is (= [[:user "some-username"] [:id "1905065-Travel-Icons-pack"] [:offset "1"] [:type "users"] [:paragraph "paragraph=3"]] inner-result)))
    (let [inner-result (recognize 
                          (pattern-info "host(dribbble.com); queryparam(offset=?offset); queryparam(list=?type); fragment(?paragraph)") 
                          "https://dribbble.com/some-username/status/1905065-Travel-Icons-pack?list=users&offset=1&page=34#paragraph=3")]
      (is (= [[:offset "1"] [:type "users"] [:paragraph "paragraph=3"]] inner-result)))

    (let [inner-result (recognize 
                          (pattern-info "host(dribbble.com); path(?user/status/?id); queryparam(list=?type); fragment(?paragraph)") 
                          "https://dribbble.com/some-username/status/1905065-Travel-Icons-pack?list=users&offset=1&page=34#paragraph=3")]
      (is (= [[:user "some-username"] [:id "1905065-Travel-Icons-pack"] [:type "users"] [:paragraph "paragraph=3"]] inner-result)))
    (let [inner-result (recognize 
                          (pattern-info "host(dribbble.com); path(?user/status/?id); fragment(?paragraph)") 
                          "https://dribbble.com/some-username/status/1905065-Travel-Icons-pack?list=users&offset=1&page=34#paragraph=3")]
      (is (= [[:user "some-username"] [:id "1905065-Travel-Icons-pack"] [:paragraph "paragraph=3"]] inner-result)))
    (let [inner-result (recognize 
                          (pattern-info "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type);") 
                          "https://dribbble.com/some-username/status/1905065-Travel-Icons-pack?list=users&offset=1&page=34#paragraph=3")]
      (is (= [[:user "some-username"] [:id "1905065-Travel-Icons-pack"] [:offset "1"] [:type "users"]] inner-result)))
    (let [inner-result (recognize 
                          (pattern-info "host(dribbble.com);") 
                          "https://dribbble.com/some-username/status/1905065-Travel-Icons-pack?list=users&offset=1&page=34#paragraph=3")]
      (is (= [] inner-result)))
    (let [inner-result (recognize 
                          (pattern-info "host();") 
                          "https://dribbble.com/some-username/status/1905065-Travel-Icons-pack?list=users&offset=1&page=34#paragraph=3")]
      (is (= [] inner-result)))
  )
)


src/dribbble_experiments/core.clj
(ns dribbble-experiments.core
  (:gen-class)
  (:require [clj-http.client :as client]
            [clojure.string :as st]
            [clojure.data.json :as json]))

(def endpoint "https://api.dribbble.com/v1/")
(def test-user "louiethelowe")
(def test-access-token "")

(defonce app-state (atom {:requests-count 0}))

(defn dr-api [path token]
  (do
    (Thread/sleep 1000)
    (swap! app-state assoc :requests-count (inc (get (deref app-state) :requests-count)))
    (println (str "<< Request #" (:requests-count (deref app-state)) " >>"))
    (json/read-str (:body (client/get (st/join "" [endpoint path]) {:insecure? true
                                                                    :force-redirects true
                                                                    :headers {"Authorization" (st/join "" ["Bearer " token])}})))))

(defn username-and-name-by-user-id [user-id token]
  (let [user-info (dr-api (st/join "" ["users/" user-id]) token)]
    [(get user-info "username") (get user-info "name")]))

(defn followers-ids [user-id token]
  (let [user-followers (dr-api (st/join "" ["users/" user-id "/followers"]) token)]
    (vec (map (fn [follower-info] (get (get follower-info "follower") "id")) user-followers))))

(defn shots-ids [user-id token]
  (let [user-shots (dr-api (st/join "" ["users/" user-id "/shots"]) token)]
    (vec (map (fn [shot-info] (get shot-info "id")) user-shots))))

(defn shots-likers-ids [shot-id token]
  (let [shot-likers (dr-api (st/join "" ["shots/" shot-id "/likes"]) token)]
    (vec (map (fn [liker-info] (get (get liker-info "user") "id")) shot-likers))))

(defn find-top-likers-ids [user-name number-of-top-likers token]
  (let [user-info (dr-api (st/join "" ["users/" user-name]) token)
        user-id (get user-info "id")
        user-followers (followers-ids user-id token)
        all-shots-ids (flatten (vec (map (fn [follower-id] (shots-ids follower-id token)) user-followers)))
        all-likers-ids-flat (remove nil? (flatten (map (fn [shot-id] (shots-likers-ids shot-id token)) all-shots-ids)))
        likes-per-user-id (frequencies all-likers-ids-flat)
        top-likers-ids (take number-of-top-likers (sort-by val > likes-per-user-id))
        ]
    (println "")
    (println "User Info:")
    (println user-info)
    (println "")
    (println "Followers of the User:")
    (println user-followers)
    (println "")
    (println "Shots made by the Followers :")
    (println all-shots-ids)
    (println "")
    (println "User IDs of the all users who liked Shots:")
    (println all-likers-ids-flat)
    (println "")
    (println "Number of Likes per user ID of the all users who liked Shots:")
    (println likes-per-user-id)
    (println "")
    (println (str "User IDs and number of Likes of the Top " number-of-top-likers " Likers:"))
    (println top-likers-ids)
    top-likers-ids))

(defn find-top-likers-names [user-name number-of-top-likers token]
  (let [top-likers-ids (find-top-likers-ids user-name number-of-top-likers token)
        top-likers-names (vec (map (fn [user-id-count] (username-and-name-by-user-id (get user-id-count 0) token)) (seq top-likers-ids)))]
    (println "")
    (println "Names of the Top Likers:")
    (println top-likers-names)
    top-likers-names))

(defn print-top-likers [user-name number-of-top-likers token]
  (let [top-likers-names (find-top-likers-names user-name number-of-top-likers token)]
    (println "")
    (println "RESULT:")
    (println "Top likers (<username> - <name>):")
    (doseq  [liker top-likers-names] 
            (do
              (println (str "\"" (get liker 0) "\" - \"" (get liker 1) "\""))))
    ))

(defn -main
  [& args]
  (let [vec-args (vec args)
        args-number (count vec-args)
        user-name (if (<= 1 args-number) (get vec-args 0))
        access-token (if (<= 2 args-number) (get vec-args 1))]
    (if (= 1 args-number) (if (= "help" (get vec-args 0)) (do
        (println "")
        (println "Usage:")
        (println "lein run <Username> <AccessToken>")
        (println "")
        (println "Example:")
        (println "lein run \"louiethelowe\" \"a5c9da88ad44ffce697b62d3a264f648848ab2031eeea91814ec9b75e44b8320\"")
        )))
    (if  (<= 2 args-number) (print-top-likers user-name 10 access-token))
    (println "")
    (println (str "DONE. Made " (:requests-count (deref app-state)) " requests."))))


src/dribbble_experiments/url_match.clj
(ns dribbble-experiments.url_match
  (:gen-class)
  (:require [clojure.string :as st]
            [clojure.data.json :as json]))

;;
;; Matcher should recognize and destruct URL by:
;; host: domain
;; path: parts, splitted with "/"
;; queryparam: name/value pairs of query
;; (see examples below)
;; 
;; Each string that is started from "?" is a "bind"
;; (recognize matcher) should return nil or seq of binds
;;

; ======================================================================================================================
; === TASK EXAMPLES ====================================================================================================

; (def twitter (Pattern. "host(twitter.com); path(?user/status/?id);"))
; (recognize twitter "http://twitter.com/bradfitz/status/562360748727611392")
; ;; => [[:id 562360748727611392] [:user "bradfitz"]]

; (def dribbble (Pattern. "host(dribbble.com); path(shots/?id); queryparam(offset=?offset);")
; (recognize dribbble "https://dribbble.com/shots/1905065-Travel-Icons-pack?list=users&offset=1")
; ;; => [[:id "1905065-Travel-Icons-pack"] [:offset "1"]]
; (recognize dribbble "https://twitter.com/shots/1905065-Travel-Icons-pack?list=users&offset=1")
; ;; => nil ;; host mismatch
; (recognize dribbble "https://dribbble.com/shots/1905065-Travel-Icons-pack?list=users")
; ;; => nil ;; offset queryparam missing

; (def dribbble2 (Pattern. "host(dribbble.com); path(shots/?id); queryparam(offset=?offset); queryparam(list=?type);")

; ======================================================================================================================
; === IMPLEMENTATION ===================================================================================================

; ==============================================================================
; === GENERIC ==================================================================

; (defn uri-allowed-chars []
;   "\\w\\-\\~\\%\\.")

(defonce uri-allowed-chars "\\w\\-\\~\\%\\.")

; :type
; from
; "list=?type"
;
; :user
; from
; "?user"
(defn param-to-keyword [pattern]
  (keyword (re-find (re-matcher (re-pattern (str "(?<=\\?)[" uri-allowed-chars "]+(?=$)")) pattern))))

; "https\\:\\/\\/dribbble\\.com\\/shots\\/1905065\\-Travel\\-Icons\\-pack\\?list\\=users"
; from
; "https://dribbble.com/shots/1905065-Travel-Icons-pack?list=users"
;
; "my\\-domain\\.com\\.eu"
; from
; "my-domain.com.eu"
(defn prepare-string-to-re-pattern [original-string]
  (st/replace original-string #"\W" #(str "\\" %1)))

; ==============================================================================
; === HOST =====================================================================

; "dribbble.com"
; from
; "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type);"
;
; ""
; from
; "host(); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type);"
;
; nil
; from
; "path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type);"
(defn host-name [pattern]
  (let [inner-empty-host? (some? (re-find #"host\(\)" pattern))]
    (if inner-empty-host? 
        ""
        (re-find (re-matcher (re-pattern (str "(?<=host\\()[" uri-allowed-chars "]+(?=\\))")) pattern)))))

; #"^(http|https)://dribbble\.com($|\/$|\/.+|\#$|\#.+)"
; from
; "dribbble.com"
;
; #"^(http|https)://[\w\-\~\%\.]+($|\/$|\/.+|\#$|\#.+)"
; from
; ""
(defn host-pattern [host-name]
  (if (some? host-name) 
      (if (> (count host-name) 0) 
          (re-pattern (str "^(http|https)://" (prepare-string-to-re-pattern host-name) "($|\\/$|\\/.+|\\#$|\\#.+)"))
          (re-pattern (str "^(http|https)://[" uri-allowed-chars "]+($|\\/$|\\/.+|\\#$|\\#.+)")))
      ))

; #"^(http|https)://dribbble\.com($|\/$|\/.+|\#$|\#.+)"
; from
; "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type);"
(defn host-info [pattern]
  (let [inner-host-name (host-name pattern)]
    (if (some? inner-host-name) (host-pattern inner-host-name))))

; true
; from
; (pattern-info "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type);")
; "https://dribbble.com/shots/1905065-Travel-Icons-pack?list=users&offset=1"
;
; false
; from
; (pattern-info "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type);")
; "https://twitter.com/shots/1905065-Travel-Icons-pack?list=users&offset=1"
(defn check-host-by-info [pattern uri]
  (some? (re-matches pattern uri)))

; ==============================================================================
; === PATH =====================================================================

; "?user/status/?id"
; from
; "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type);"
(defn path [pattern]
  (re-find (re-matcher (re-pattern (str "(?<=path\\()[" uri-allowed-chars "\\?\\/]+(?=\\))")) pattern)))

; ["?user" "status" "?id"]
; from
; "?user/status/?id"
(defn path-to-pieces [inner-path-pattern]
  (st/split inner-path-pattern #"/"))

; "status"
; from
; "status"
;
; nil
; from
; "?status"
(defn path-piece-name [path-piece]
  (re-matches (re-pattern (str "[" uri-allowed-chars "]+")) path-piece))

; [[0 :user nil] [1 nil "status"] [2 :id nil]]
; from
; "?user/status/?id"
(defn path-param-info [inner-path-pattern]
  (let [path-pieces (path-to-pieces inner-path-pattern)]
    (vec (map-indexed (fn [num item] [num (param-to-keyword item) (path-piece-name item)]) path-pieces))))

; #"(?<=^(?:http|https)://dribbble\.com\/)[\w\-\~\%\.\/]+(?=\?)"
; from
; "dribbble.com"
;
; #"(?<=^(?:http|https)://[\w\-\~\%\.]+\/)[\w\-\~\%\.\/]+(?=\?)"
; from
; "dribbble.com"
(defn path-pattern [host-name]
  (if (some? host-name) 
      (if (> (count host-name) 0)
          (re-pattern (str "(?<=^(?:http|https)://" (prepare-string-to-re-pattern host-name) "\\/)[" uri-allowed-chars "\\/]+(?=\\?)")))))

(defonce path-prefix-pattern #"^(?:http|https)://[\w\-\~\%\.]+(?=.+)")

(defonce path-postfix-pattern (re-pattern (str "(?<=\\/)[" uri-allowed-chars "\\/]+(?=\\?)")))

; {:host-was-defined true, :path-pattern #"(?<=^(?:http|https)://dribbble\.com\/)[\w\-\~\%\.\/]+(?=\?)", :path-pieces [[0 :user nil] [1 nil "status"] [2 :id nil]]}
; from
; "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type);"
(defn path-info [pattern]
  (let [inner-path (path pattern)
        inner-host-name (host-name pattern)
        inner-host-name-defined? (if (some? inner-host-name) (if (> (count inner-host-name) 0) true false))
        inner-path-pattern (if (some? inner-path) (path-pattern inner-host-name))
        inner-path-prefix-pattern ()]
    (if (some? inner-path) {
      :host-was-defined inner-host-name-defined?
      :path-pattern inner-path-pattern
      :path-pieces (path-param-info inner-path)})))

; "shots/1905065-Travel-Icons-pack"
; from
; "https://dribbble.com/shots/1905065-Travel-Icons-pack?list=users&offset=1"
; #"(?<=^(?:http|https)://dribbble\.com\/)[\w\-\~\%\.\/]+(?=\?)"
(defn get-path-with-info [uri pattern]
  (re-find (re-matcher pattern uri)))

; "shots/1905065-Travel-Icons-pack"
; from
; "https://dribbble.com/shots/1905065-Travel-Icons-pack?list=users&offset=1"
; #"(?<=^(?:http|https)://dribbble\.com\/)[\w\-\~\%\.\/]+(?=\?)"
(defn get-path-with-info-for-undefined-host [uri]
  (let [inner-host-part (re-find (re-matcher path-prefix-pattern uri))
        inner-host-part-size (count inner-host-part)
        inner-post-host-part (subs uri inner-host-part-size)]
    (re-find (re-matcher path-postfix-pattern inner-post-host-part))))

; [[:user "some-username"] [:id "1905065-Travel-Icons-pack"]]
; from
; "https://dribbble.com/some-username/status/1905065-Travel-Icons-pack?list=users&offset=1"
; {:host-was-defined true, :path-pattern #"(?<=^(?:http|https)://dribbble\.com\/)[\w\-\~\%\.\/]+(?=\?)", :path-pieces [[0 :user nil] [1 nil "status"] [2 :id nil]]}
(defn generate-path-answer [uri inner-path-info]
  (let [can-be? (some? inner-path-info)
        path-pattern-good (if can-be? (some? (:host-was-defined inner-path-info)))
        path-pattern-defined (if path-pattern-good (:host-was-defined inner-path-info))
        inner-relative-part (if path-pattern-good (if path-pattern-defined 
                                                      (get-path-with-info uri (:path-pattern inner-path-info))
                                                      (get-path-with-info-for-undefined-host uri)))
        inner-path-pieces (if (some? inner-relative-part) (path-to-pieces inner-relative-part))
        inner-pieces-info (if can-be? (:path-pieces inner-path-info))
        inner-is-enough-pieces? (if can-be? (>= (count inner-path-pieces) (count inner-pieces-info)) false)
        inner-is-good-to-go? (and can-be? inner-is-enough-pieces?)]
    (if can-be? 
      (if inner-is-enough-pieces?
        (remove nil? (vec (map #((fn [path-pieces piece-info] 
            (if (some? (get piece-info 1)) 
              [
                (get piece-info 1) 
                (get path-pieces (get piece-info 0))
              ]
              (if (not (= (get piece-info 2) (get path-pieces (get piece-info 0))))
                [nil nil])
            )
          ) inner-path-pieces %) inner-pieces-info)))
        [nil]))))

; ==============================================================================
; === QUERYPARAMS ==============================================================

; ["offset=?offset" "list=?type"]
; from
; "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type);"
(defn queryparams [pattern]
  (let [inner-result (re-seq (re-pattern (str "(?<=queryparam\\()[" uri-allowed-chars "\\?\\=\\/]+(?=\\))")) pattern)]
    (if (some? inner-result) (vec inner-result))))

; "list"
; from
; "list=?type"
(defn query-param-name [pattern]
  (re-find (re-matcher (re-pattern (str "(?<=^)[" uri-allowed-chars "]+(?=\\=\\?)")) pattern)))

; #"(?<=[\?\&]list\=)[\w\-\%]+(?=[$\&\#])"
; from
; "list"
(defn query-param-pattern [q-param-name]
  (if (some? q-param-name) (re-pattern (str "(?<=[\\?\\&]" q-param-name "\\=)[\\w\\-\\%]+(?=[$\\&\\#])"))))

; [:type #"(?<=[\?\&]list\=)[\w\-\%]+(?=[$\&\#])"]
; from
; "list=?type"
(defn query-param-info [pattern]
  (let [inner-keyword (param-to-keyword pattern)
        inner-query-param-pattern (query-param-pattern (query-param-name pattern))
        inner-good-to-go? (and (some? inner-keyword) (some? inner-query-param-pattern))]
    (if inner-good-to-go? [inner-keyword inner-query-param-pattern])))

; [[:offset #"(?<=[\?\&]offset\=)[\w\-\%]+(?=[$\&\#])"] [:type #"(?<=[\?\&]list\=)[\w\-\%]+(?=[$\&\#])"]]
; from
; "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type);"
(defn query-info [pattern]
  (let [inner-queryparams (queryparams pattern)
        inner-result (if (some? inner-queryparams) (remove nil? (vec (map query-param-info inner-queryparams))))]
    (if (some? inner-result)
        (if (> (count inner-result) 0) inner-result))))

; "users"
; from
; "https://twitter.com/shots/1905065-Travel-Icons-pack?list=users&offset=1"
; #"(?<=[\?\&]list\=)[\w\-\%]+(?=[$\&\#])"
(defn get-query-param-with-info [uri pattern]
  (re-find (re-matcher pattern uri)))

; [[:list "users"] [:offset "1"]]
; from
; "https://twitter.com/shots/1905065-Travel-Icons-pack?list=users&offset=1#some-my/fragment"
; [[:offset #"(?<=[\?\&]offset\=)[\w\-\%]+(?=[$\&\#])"] [:type #"(?<=[\?\&]list\=)[\w\-\%]+(?=[$\&\#])"]]
(defn generate-query-answer [uri inner-query-info]
  (if (some? inner-query-info)
    (vec (map #((fn [uri param-info] [(get param-info 0) (get-query-param-with-info uri (get param-info 1))]) uri %) inner-query-info))))

; ==============================================================================
; === FRAGMENT ==============================================================

; "?paragraph"
; from
; "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type); fragment(?paragraph)"
(defn fragment [pattern]
  (re-find (re-matcher (re-pattern (str "(?<=fragment\\()[" uri-allowed-chars "\\?]+(?=\\))")) pattern)))

; #"(?<=\#).+(?=$)"
(defn fragment-pattern []
  #"(?<=\#).+(?=$)")

; [:paragraph #"(?<=\#).+(?=$)"]
; from
; "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type); fragment(?paragraph)"
(defn fragment-info [pattern]
  (let [inner-fragment (fragment pattern)
        inner-keyword (if (some? inner-fragment) (param-to-keyword inner-fragment))
        inner-good-to-go? (and (some? inner-fragment) (some? inner-keyword))]
    (if inner-good-to-go? [inner-keyword (fragment-pattern)])))

; "some-my/fragment"
; from
; "https://twitter.com/shots/1905065-Travel-Icons-pack?list=users&offset=1#some-my/fragment"
; #"(?<=\#).+(?=$)"
(defn get-fragment-with-info [uri pattern]
  (re-find (re-matcher pattern uri)))

; [[:paragraph "some-my/fragment"]]
; from
; "https://twitter.com/shots/1905065-Travel-Icons-pack?list=users&offset=1#some-my/fragment"
; [:paragraph #"(?<=\#).+(?=$)"]
(defn generate-fragment-answer [uri inner-fragment-info]
  (if (some? inner-fragment-info)
    [[(get inner-fragment-info 0) (get-fragment-with-info uri (get inner-fragment-info 1))]]))

; ==============================================================================
; === API ======================================================================

; {:host #"^(http|https)://dribbble\.com($|\/$|\/.+|\#$|\#.+)", :path {:path-pattern #"(?<=^(?:http|https)://dribbble\.com\/)[\w\-\~\%\.\/]+(?=\?)", :path-pieces [[0 :user nil] [1 nil "status"] [2 :id nil]]}, :query [[:offset #"(?<=[\?\&]offset\=)[\w\-\%]+(?=[$\&\#])"] [:type #"(?<=[\?\&]list\=)[\w\-\%]+(?=[$\&\#])"]], :fragment [:paragraph #"(?<=\#).+(?=$)"]}
; from
; "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type); fragment(?paragraph)"
(defn pattern-info [pattern]
  { :host (host-info pattern)
    :path (path-info pattern)
    :query (query-info pattern)
    :fragment (fragment-info pattern)})

; [[:user "some-username"] [:id "1905065-Travel-Icons-pack"] [:offset "1"] [:type "users"] [:paragraph "paragraph=3"]]
; from
; (pattern-info "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type); fragment(?paragraph)")
; "https://dribbble.com/some-username/status/1905065-Travel-Icons-pack?list=users&offset=1&page=34#paragraph=3"
;
; nil
; from
; (pattern-info "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type); fragment(?paragraph)")
; "https://twitter.com/some-username/status/1905065-Travel-Icons-pack?list=users&offset=1&page=34#paragraph=3"
;
; [[:user "some-username"] [nil nil] [:id "weight"] [:offset "1"] [:type nil] [:paragraph nil]]
; from
; (pattern-info "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type); fragment(?paragraph)")
; "https://dribbble.com/some-username/height/weight/1905065-Travel-Icons-pack?listing=users&offset=1&page=34"
(defn recognize-detailed [pattern uri]
  (let [inner-host-info (:host pattern)
        is-host-ok? (if (some? inner-host-info)
                        (check-host-by-info inner-host-info uri)
                        true)
        inner-path-info (:path pattern)
        inner-query-info (:query pattern)
        inner-fragment-info (:fragment pattern)]
    (if is-host-ok?
      (vec (concat
        (generate-path-answer uri inner-path-info)
        (generate-query-answer uri inner-query-info)
        (generate-fragment-answer uri inner-fragment-info))))))

; [[:user "some-username"] [:id "1905065-Travel-Icons-pack"] [:offset "1"] [:type "users"] [:paragraph "paragraph=3"]]
; from
; (pattern-info "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type); fragment(?paragraph)")
; "https://dribbble.com/some-username/status/1905065-Travel-Icons-pack?list=users&offset=1&page=34#paragraph=3"
;
; nil
; from
; (pattern-info "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type); fragment(?paragraph)")
; "https://twitter.com/some-username/status/1905065-Travel-Icons-pack?list=users&offset=1&page=34#paragraph=3"
;
; nil
; from
; (pattern-info "host(dribbble.com); path(?user/status/?id); queryparam(offset=?offset); queryparam(list=?type); fragment(?paragraph)")
; "https://dribbble.com/some-username/height/weight/1905065-Travel-Icons-pack?listing=users&offset=1&page=34"
(defn recognize [pattern uri]
  (let [detailed-result (recognize-detailed pattern uri)]
    (if (some? detailed-result)
        (if (not (contains? (set (flatten detailed-result)) nil))
            detailed-result))))
